indigo_app/static/javascript/indigo/models.js

```diff
@@ -40,27 +45,86 @@
       return this.document.url() + '/content';
     },
 
+    setupMutationObserver: function () {
+      this.observer = new MutationObserver((mutations) => {
+        for (const mutation of mutations) {
+          console.log('mutation', mutation);
+          this.trigger('mutation', this, mutation);
+        }
+        this.trigger('change:dom', this);
+        this.trigger('change', this);
+      });
+
+      this.observer.observe(this.xmlDocument, {
+        childList: true,
+        attributes: true,
+        subtree: true,
+      });
+    },
+
+    /**
+     * Determine the impact of a mutation on the provided element.
+     * @param mutation MutationRecord
+     * @param element Element in this XML document
+     * @returns 'changed' if the mutation impacts the element, 'removed' if the element was removed from the tree,
+     *          'replaced', if the element has been replaced (with a node in mutation.addedNodes),
+     *          or null if there is no impact
+     */
+    getMutationImpact(mutation, element) {
+      const target = mutation.target;
+
+      if (mutation.type === 'childList') {
+        if (mutation.removedNodes[0] === element) {
+          // the element has been replaced
+          return 'replaced';
+        }
+
+        const ownerDocument = target.nodeType === Node.DOCUMENT_NODE ? target : target.ownerDocument;
+        if (!ownerDocument.contains(element)) {
+          // the change removed xmlElement from the tree
+          return 'removed';
```

Comment:
This is returned when the editor is open, a change has been made but not saved, and some document analysis is run:
- The editor is then closed and we scroll to the top of the selected ToC element or quick-edited element
- The italics terms or whatever are applied
- AND the change made in the editor is kept, but this isn't easy to see or obvious if the quick-edited provision is out of sight.

Probably we should discard changes in the editor first (with confirmation), or disable the analysis menu while the editor is open. In the current world, the editor is closed and the unsaved changes are lost, but the user isn't warned.

Reply:
Hmm, good point. Analysis replaces the whole document. Any analyser view that does that should first call:

```
if (Indigo.view.sourceEditorView.confirmAndDiscardChanges()) {
...
}
```

Reply:
I'll make this change

Reply:
done